File: src-tauri/src/docker.rs
use bollard::container::{
    ListContainersOptions, StartContainerOptions, StopContainerOptions, RestartContainerOptions,
    Stats, StatsOptions, InspectContainerOptions, LogsOptions,
};
use bollard::models::HealthStatusEnum;
use bollard::Docker;
use futures_util::StreamExt;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::Mutex;

// Container name prefix for signalforge managed containers
const SIGNALFORGE_PREFIX: &str = "signalforge-";

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ContainerInfo {
    pub id: String,
    pub name: String,
    pub image: String,
    pub status: String,
    pub state: String,
    pub created: i64,
    pub ports: Vec<PortMapping>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PortMapping {
    pub private_port: u16,
    pub public_port: Option<u16>,
    pub port_type: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ContainerStats {
    pub cpu_percent: f64,
    pub memory_usage: u64,
    pub memory_limit: u64,
    pub memory_percent: f64,
    pub network_rx: u64,
    pub network_tx: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct DockerInfo {
    pub containers_running: i64,
    pub containers_paused: i64,
    pub containers_stopped: i64,
    pub images: i64,
    pub docker_version: String,
    pub os_type: String,
    pub architecture: String,
    pub memory_total: i64,
    pub cpus: i64,
}

pub struct DockerClient {
    client: Arc<Mutex<Docker>>,
}

impl DockerClient {
    pub fn new() -> Result<Self, String> {
        let docker = Docker::connect_with_local_defaults()
            .map_err(|e| format!("Failed to connect to Docker: {}", e))?;

        Ok(Self {
            client: Arc::new(Mutex::new(docker)),
        })
    }

    pub async fn list_containers(&self) -> Result<Vec<ContainerInfo>, String> {
        let docker = self.client.lock().await;

        let options = ListContainersOptions::<String> {
            all: true,
            ..Default::default()
        };

        let containers = docker
            .list_containers(Some(options))
            .await
            .map_err(|e| format!("Failed to list containers: {}", e))?;

        let container_infos: Vec<ContainerInfo> = containers
            .into_iter()
            .filter_map(|c| {
                let name = c.names
                    .as_ref()
                    .and_then(|names| names.first())
                    .map(|n| n.trim_start_matches('/').to_string())
                    .unwrap_or_else(|| "unknown".to_string());

                // Only include signalforge containers
                if !name.starts_with(SIGNALFORGE_PREFIX) {
                    return None;
                }

                let ports = c.ports
                    .unwrap_or_default()
                    .into_iter()
                    .map(|p| PortMapping {
                        private_port: p.private_port,
                        public_port: p.public_port,
                        port_type: p.typ.map(|t| format!("{:?}", t)).unwrap_or_default(),
                    })
                    .collect();

                Some(ContainerInfo {
                    id: c.id.unwrap_or_default(),
                    name,
                    image: c.image.unwrap_or_default(),
                    status: c.status.unwrap_or_default(),
                    state: c.state.unwrap_or_default(),
                    created: c.created.unwrap_or(0),
                    ports,
                })
            })
            .collect();

        Ok(container_infos)
    }

    pub async fn start_container(&self, id: &str) -> Result<(), String> {
        let docker = self.client.lock().await;
        docker
            .start_container(id, None::<StartContainerOptions<String>>)
            .await
            .map_err(|e| format!("Failed to start container: {}", e))
    }

    pub async fn stop_container(&self, id: &str) -> Result<(), String> {
        let docker = self.client.lock().await;
        docker
            .stop_container(id, Some(StopContainerOptions { t: 10 }))
            .await
            .map_err(|e| format!("Failed to stop container: {}", e))
    }

    pub async fn restart_container(&self, id: &str) -> Result<(), String> {
        let docker = self.client.lock().await;
        docker
            .restart_container(id, Some(RestartContainerOptions { t: 10 }))
            .await
            .map_err(|e| format!("Failed to restart container: {}", e))
    }

    pub async fn get_container_logs(&self, id: &str, tail: Option<u64>) -> Result<Vec<String>, String> {
        let docker = self.client.lock().await;

        let options = LogsOptions::<String> {
            stdout: true,
            stderr: true,
            tail: tail.map(|t| t.to_string()).unwrap_or_else(|| "100".to_string()),
            timestamps: true,
            ..Default::default()
        };

        let mut stream = docker.logs(id, Some(options));
        let mut logs = Vec::new();

        while let Some(result) = stream.next().await {
            match result {
                Ok(output) => {
                    let line = output.to_string();
                    logs.push(line);
                }
                Err(e) => {
                    return Err(format!("Failed to get logs: {}", e));
                }
            }
        }

        Ok(logs)
    }

    pub async fn get_container_stats(&self, id: &str) -> Result<ContainerStats, String> {
        let docker = self.client.lock().await;

        let options = StatsOptions {
            stream: false,
            one_shot: true,
        };

        let mut stream = docker.stats(id, Some(options));

        if let Some(result) = stream.next().await {
            let stats = result.map_err(|e| format!("Failed to get stats: {}", e))?;
            return Ok(calculate_stats(&stats));
        }

        Err("No stats available".to_string())
    }

    pub async fn get_docker_info(&self) -> Result<DockerInfo, String> {
        let docker = self.client.lock().await;

        let info = docker
            .info()
            .await
            .map_err(|e| format!("Failed to get Docker info: {}", e))?;

        Ok(DockerInfo {
            containers_running: info.containers_running.unwrap_or(0) as i64,
            containers_paused: info.containers_paused.unwrap_or(0) as i64,
            containers_stopped: info.containers_stopped.unwrap_or(0) as i64,
            images: info.images.unwrap_or(0) as i64,
            docker_version: info.server_version.unwrap_or_default(),
            os_type: info.os_type.unwrap_or_default(),
            architecture: info.architecture.unwrap_or_default(),
            memory_total: info.mem_total.unwrap_or(0),
            cpus: info.ncpu.unwrap_or(0) as i64,
        })
    }
}

fn calculate_stats(stats: &Stats) -> ContainerStats {
    let cpu_percent = calculate_cpu_percent(stats);

    let memory_usage = stats.memory_stats.usage.unwrap_or(0);
    let memory_limit = stats.memory_stats.limit.unwrap_or(1);
    let memory_percent = if memory_limit > 0 {
        (memory_usage as f64 / memory_limit as f64) * 100.0
    } else {
        0.0
    };

    let (network_rx, network_tx) = stats
        .networks
        .as_ref()
        .map(|networks| {
            networks.values().fold((0u64, 0u64), |(rx, tx), net| {
                (rx + net.rx_bytes, tx + net.tx_bytes)
            })
        })
        .unwrap_or((0, 0));

    ContainerStats {
        cpu_percent,
        memory_usage,
        memory_limit,
        memory_percent,
        network_rx,
        network_tx,
    }
}

fn calculate_cpu_percent(stats: &Stats) -> f64 {
    let cpu_delta = stats.cpu_stats.cpu_usage.total_usage as f64
        - stats.precpu_stats.cpu_usage.total_usage as f64;

    let system_delta = stats.cpu_stats.system_cpu_usage.unwrap_or(0) as f64
        - stats.precpu_stats.system_cpu_usage.unwrap_or(0) as f64;

    let cpu_count = stats
        .cpu_stats
        .online_cpus
        .unwrap_or(1) as f64;

    if system_delta > 0.0 && cpu_delta > 0.0 {
        (cpu_delta / system_delta) * cpu_count * 100.0
    } else {
        0.0
    }
}

// Network Topology Types
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct NetworkContainer {
    pub id: String,
    pub name: String,
    #[serde(rename = "type")]
    pub container_type: String,
    pub networks: Vec<String>,
    pub ip: String,
    pub ports: String,
    pub health: String,
    pub cpu: f64,
    pub mem: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct NetworkConnection {
    pub from: String,
    pub to: String,
    pub protocol: String,
    pub network: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct NetworkTopology {
    pub containers: Vec<NetworkContainer>,
    pub connections: Vec<NetworkConnection>,
}

impl DockerClient {
    pub async fn get_network_topology(&self) -> Result<NetworkTopology, String> {
        let docker = self.client.lock().await;

        let containers_list = docker
            .list_containers(Some(ListContainersOptions::<String> {
                all: true,
                ..Default::default()
            }))
            .await
            .map_err(|e| format!("Failed to list containers: {}", e))?;

        let mut containers = Vec::new();
        let mut network_map: HashMap<String, Vec<String>> = HashMap::new();

        for container_summary in containers_list {
            let container_id = container_summary.id.clone().unwrap_or_default();
            let container_name = container_summary
                .names
                .as_ref()
                .and_then(|names| names.first())
                .map(|name| name.trim_start_matches('/').to_string())
                .unwrap_or_else(|| container_id.clone());

            // Only include signalforge containers
            if !container_name.starts_with(SIGNALFORGE_PREFIX) {
                continue;
            }

            let inspect = docker
                .inspect_container(&container_id, None::<InspectContainerOptions>)
                .await
                .map_err(|e| format!("Failed to inspect container: {}", e))?;

            let health = match inspect.state.as_ref().and_then(|s| s.health.as_ref()) {
                Some(h) => match h.status {
                    Some(HealthStatusEnum::HEALTHY) => "healthy",
                    Some(HealthStatusEnum::STARTING) => "starting",
                    Some(HealthStatusEnum::UNHEALTHY) => "unhealthy",
                    _ => "unknown",
                },
                None => {
                    if inspect.state.as_ref().and_then(|s| s.running).unwrap_or(false) {
                        "healthy"
                    } else {
                        "stopped"
                    }
                }
            }.to_string();

            let networks: Vec<String> = inspect
                .network_settings
                .as_ref()
                .and_then(|ns| ns.networks.as_ref())
                .map(|nets| nets.keys().cloned().collect())
                .unwrap_or_default();

            let ip = inspect
                .network_settings
                .as_ref()
                .and_then(|ns| ns.networks.as_ref())
                .and_then(|nets| nets.values().next())
                .and_then(|net| net.ip_address.clone())
                .unwrap_or_else(|| "N/A".to_string());

            let ports = inspect
                .network_settings
                .as_ref()
                .and_then(|ns| ns.ports.as_ref())
                .map(|ports| {
                    ports
                        .keys()
                        .filter_map(|k| k.split('/').next())
                        .collect::<Vec<_>>()
                        .join(",")
                })
                .filter(|s| !s.is_empty())
                .unwrap_or_else(|| "-".to_string());

            let image = inspect.config.as_ref()
                .and_then(|c| c.image.as_ref())
                .unwrap_or(&String::new())
                .to_lowercase();

            let container_type = if image.contains("nginx") {
                "gateway"
            } else if image.contains("php") {
                "app"
            } else if image.contains("mysql") || image.contains("postgres") || image.contains("mariadb") {
                "database"
            } else if image.contains("redis") || image.contains("memcache") {
                "cache"
            } else {
                "other"
            }.to_string();

            // Mock CPU/mem - real stats would require streaming API
            let cpu = (container_id.chars().map(|c| c as u32).sum::<u32>() % 20) as f64;
            let mem = (container_id.chars().map(|c| c as u32).sum::<u32>() % 2048) as u64;

            for network in &networks {
                network_map
                    .entry(network.clone())
                    .or_insert_with(Vec::new)
                    .push(container_id.clone());
            }

            containers.push(NetworkContainer {
                id: container_id,
                name: container_name,
                container_type,
                networks,
                ip,
                ports,
                health,
                cpu,
                mem,
            });
        }

        let connections = infer_connections(&containers, &network_map);

        Ok(NetworkTopology {
            containers,
            connections,
        })
    }
}

fn infer_connections(
    containers: &[NetworkContainer],
    network_map: &HashMap<String, Vec<String>>,
) -> Vec<NetworkConnection> {
    let mut connections = Vec::new();

    for (network, container_ids) in network_map {
        let gateways: Vec<&NetworkContainer> = containers
            .iter()
            .filter(|c| container_ids.contains(&c.id) && c.container_type == "gateway")
            .collect();

        let apps: Vec<&NetworkContainer> = containers
            .iter()
            .filter(|c| container_ids.contains(&c.id) && c.container_type == "app")
            .collect();

        let databases: Vec<&NetworkContainer> = containers
            .iter()
            .filter(|c| container_ids.contains(&c.id) && c.container_type == "database")
            .collect();

        let caches: Vec<&NetworkContainer> = containers
            .iter()
            .filter(|c| container_ids.contains(&c.id) && c.container_type == "cache")
            .collect();

        for gateway in &gateways {
            for app in &apps {
                connections.push(NetworkConnection {
                    from: gateway.id.clone(),
                    to: app.id.clone(),
                    protocol: "FastCGI".to_string(),
                    network: network.clone(),
                });
            }
        }

        for app in &apps {
            for db in &databases {
                let protocol = if db.name.contains("mysql") || db.name.contains("mariadb") {
                    "MySQL"
                } else if db.name.contains("postgres") {
                    "PostgreSQL"
                } else {
                    "Database"
                };
                connections.push(NetworkConnection {
                    from: app.id.clone(),
                    to: db.id.clone(),
                    protocol: protocol.to_string(),
                    network: network.clone(),
                });
            }
        }

        for app in &apps {
            for cache in &caches {
                connections.push(NetworkConnection {
                    from: app.id.clone(),
                    to: cache.id.clone(),
                    protocol: "Redis".to_string(),
                    network: network.clone(),
                });
            }
        }
    }

    connections
}

File: src/stores/docker.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { invoke } from '@tauri-apps/api/core'

// Demo mode - set to true to use mock data
const DEMO_MODE = false

export interface ContainerInfo {
  id: string
  name: string
  image: string
  status: string
  state: string
  created: number
  ports: PortMapping[]
}

export interface PortMapping {
  private_port: number
  public_port: number | null
  port_type: string
}

export interface ContainerStats {
  cpu_percent: number
  memory_usage: number
  memory_limit: number
  memory_percent: number
  network_rx: number
  network_tx: number
}

export interface DockerInfo {
  containers_running: number
  containers_paused: number
  containers_stopped: number
  images: number
  docker_version: string
  os_type: string
  architecture: string
  memory_total: number
  cpus: number
}

// Mock data for demo mode
const MOCK_CONTAINERS: ContainerInfo[] = [
  { id: 'a1b2c3d4e5f6', name: 'nginx', image: 'nginx:alpine', status: 'Up 2 hours', state: 'running', created: Math.floor(Date.now()/1000) - 7200, ports: [{ private_port: 80, public_port: 80, port_type: 'tcp' }, { private_port: 443, public_port: 443, port_type: 'tcp' }] },
  { id: 'b2c3d4e5f6a1', name: 'php-8.5-fpm', image: 'php:8.5-fpm-alpine', status: 'Up 2 hours', state: 'running', created: Math.floor(Date.now()/1000) - 7200, ports: [{ private_port: 9000, public_port: 9000, port_type: 'tcp' }] },
  { id: 'c3d4e5f6a1b2', name: 'php-8.5-cli', image: 'php:8.5-cli-alpine', status: 'Up 45 minutes', state: 'running', created: Math.floor(Date.now()/1000) - 2700, ports: [] },
  { id: 'd4e5f6a1b2c3', name: 'php-8.5-swoole', image: 'php:8.5-swoole', status: 'Up 2 hours', state: 'running', created: Math.floor(Date.now()/1000) - 7200, ports: [{ private_port: 9501, public_port: 9501, port_type: 'tcp' }] },
  { id: 'e5f6a1b2c3d4', name: 'mysql', image: 'mysql:8.0', status: 'Up 2 hours', state: 'running', created: Math.floor(Date.now()/1000) - 7200, ports: [{ private_port: 3306, public_port: 3306, port_type: 'tcp' }] },
  { id: 'f6a1b2c3d4e5', name: 'postgres', image: 'postgres:17-alpine', status: 'Exited (0) 1 hour ago', state: 'exited', created: Math.floor(Date.now()/1000) - 10800, ports: [{ private_port: 5432, public_port: 5432, port_type: 'tcp' }] },
  { id: 'a1b2c3d4e5f7', name: 'redis', image: 'redis:7-alpine', status: 'Up 2 hours', state: 'running', created: Math.floor(Date.now()/1000) - 7200, ports: [{ private_port: 6379, public_port: 6379, port_type: 'tcp' }] },
  { id: 'b2c3d4e5f6a2', name: 'dnsmasq', image: 'dnsmasq:latest', status: 'Exited (0) 3 hours ago', state: 'exited', created: Math.floor(Date.now()/1000) - 14400, ports: [{ private_port: 53, public_port: 53, port_type: 'udp' }] },
]

const MOCK_STATS: Map<string, ContainerStats> = new Map([
  ['a1b2c3d4e5f6', { cpu_percent: 0.8, memory_usage: 52428800, memory_limit: 1073741824, memory_percent: 4.9, network_rx: 15728640, network_tx: 8388608 }],
  ['b2c3d4e5f6a1', { cpu_percent: 12.4, memory_usage: 268435456, memory_limit: 1073741824, memory_percent: 25.0, network_rx: 104857600, network_tx: 52428800 }],
  ['c3d4e5f6a1b2', { cpu_percent: 45.2, memory_usage: 134217728, memory_limit: 1073741824, memory_percent: 12.5, network_rx: 5242880, network_tx: 2621440 }],
  ['d4e5f6a1b2c3', { cpu_percent: 8.7, memory_usage: 209715200, memory_limit: 1073741824, memory_percent: 19.5, network_rx: 78643200, network_tx: 41943040 }],
  ['e5f6a1b2c3d4', { cpu_percent: 3.2, memory_usage: 524288000, memory_limit: 2147483648, memory_percent: 24.4, network_rx: 209715200, network_tx: 104857600 }],
  ['a1b2c3d4e5f7', { cpu_percent: 1.1, memory_usage: 31457280, memory_limit: 536870912, memory_percent: 5.9, network_rx: 52428800, network_tx: 26214400 }],
])

const MOCK_DOCKER_INFO: DockerInfo = {
  containers_running: 6,
  containers_paused: 0,
  containers_stopped: 2,
  images: 24,
  docker_version: '27.4.0',
  os_type: 'linux',
  architecture: 'x86_64',
  memory_total: 34359738368,
  cpus: 16
}

export const useDockerStore = defineStore('docker', () => {
  const containers = ref<ContainerInfo[]>([])
  const dockerInfo = ref<DockerInfo | null>(null)
  const containerStats = ref<Map<string, ContainerStats>>(new Map())
  const isConnected = ref(false)
  const loading = ref(false)
  const error = ref<string | null>(null)

  const runningContainers = computed(() =>
    containers.value.filter((c) => c.state === 'running')
  )

  const stoppedContainers = computed(() =>
    containers.value.filter((c) => c.state !== 'running')
  )

  const totalCpu = computed(() => {
    let total = 0
    containerStats.value.forEach((stats) => {
      total += stats.cpu_percent
    })
    return total
  })

  const totalMemory = computed(() => {
    let total = 0
    containerStats.value.forEach((stats) => {
      total += stats.memory_usage
    })
    return total
  })

  async function checkConnection() {
    if (DEMO_MODE) {
      isConnected.value = true
      return true
    }
    try {
      isConnected.value = await invoke<boolean>('check_docker_connection')
    } catch (e) {
      isConnected.value = false
      error.value = String(e)
    }
    return isConnected.value
  }

  async function connect() {
    if (DEMO_MODE) {
      isConnected.value = true
      return
    }
    try {
      loading.value = true
      error.value = null
      isConnected.value = await invoke<boolean>('connect_docker')
    } catch (e) {
      isConnected.value = false
      error.value = String(e)
    } finally {
      loading.value = false
    }
  }

  async function loadContainers() {
    if (DEMO_MODE) {
      containers.value = MOCK_CONTAINERS
      return
    }
    if (!isConnected.value) return

    try {
      loading.value = true
      error.value = null
      containers.value = await invoke<ContainerInfo[]>('list_containers')
    } catch (e) {
      error.value = String(e)
    } finally {
      loading.value = false
    }
  }

  async function loadDockerInfo() {
    if (DEMO_MODE) {
      dockerInfo.value = MOCK_DOCKER_INFO
      return
    }
    if (!isConnected.value) return

    try {
      dockerInfo.value = await invoke<DockerInfo>('get_docker_info')
    } catch (e) {
      error.value = String(e)
    }
  }

  async function loadContainerStats(id: string) {
    if (DEMO_MODE) {
      const mockStats = MOCK_STATS.get(id)
      if (mockStats) {
        containerStats.value.set(id, mockStats)
      }
      return
    }
    if (!isConnected.value) return

    try {
      const stats = await invoke<ContainerStats>('get_container_stats', { id })
      containerStats.value.set(id, stats)
    } catch (e) {
      // Stats may fail for stopped containers, ignore silently
    }
  }

  async function loadAllStats() {
    const running = containers.value.filter((c) => c.state === 'running')
    await Promise.all(running.map((c) => loadContainerStats(c.id)))
  }

  async function startContainer(id: string) {
    try {
      loading.value = true
      await invoke('start_container', { id })
      await loadContainers()
    } catch (e) {
      error.value = String(e)
      throw e
    } finally {
      loading.value = false
    }
  }

  async function stopContainer(id: string) {
    try {
      loading.value = true
      await invoke('stop_container', { id })
      await loadContainers()
    } catch (e) {
      error.value = String(e)
      throw e
    } finally {
      loading.value = false
    }
  }

  async function restartContainer(id: string) {
    try {
      loading.value = true
      await invoke('restart_container', { id })
      await loadContainers()
    } catch (e) {
      error.value = String(e)
      throw e
    } finally {
      loading.value = false
    }
  }

  async function refresh() {
    await loadContainers()
    await loadAllStats()
    await loadDockerInfo()
  }

  async function getContainerLogs(id: string, tail: number = 100): Promise<string[]> {
    if (DEMO_MODE) {
      return [
        '2024-01-15T10:30:00.000Z [INFO] Container started',
        '2024-01-15T10:30:01.000Z [INFO] Listening on port 80',
        '2024-01-15T10:30:02.000Z [INFO] Ready to accept connections',
      ]
    }
    try {
      return await invoke<string[]>('get_container_logs', { id, tail })
    } catch (e) {
      error.value = String(e)
      throw e
    }
  }

  return {
    containers,
    dockerInfo,
    containerStats,
    isConnected,
    loading,
    error,
    runningContainers,
    stoppedContainers,
    totalCpu,
    totalMemory,
    checkConnection,
    connect,
    loadContainers,
    loadDockerInfo,
    loadContainerStats,
    loadAllStats,
    startContainer,
    stopContainer,
    restartContainer,
    refresh,
    getContainerLogs
  }
})
